ldif/marshal.go

```diff
@@ -0,0 +1,273 @@
+package ldif
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"gopkg.in/ldap.v2"
+	"io"
+)
+
+var foldWidth = 76
+
+// ErrMixed is the error, that we cannot mix change records and content
+// records in one LDIF
+var ErrMixed = errors.New("cannot mix change records and content records")
+
+// Marshal returns an LDIF string from the given LDIF.
+//
+// The default line lenght is 76 characters. This can be changed by setting
+// the fw parameter to something else than 0.
+// For a fold width < 0, no folding will be done, with 0, the default is used.
+func Marshal(l *LDIF) (data string, err error) {
+	hasEntry := false
+	hasChange := false
+
+	if l.Version > 0 {
+		data = "version: 1\n"
+	}
+
+	fw := l.FoldWidth
+	if fw == 0 {
+		fw = foldWidth
+	}
+
+	for _, e := range l.Entries {
+		switch {
+		case e.Add != nil:
+			hasChange = true
+			if hasEntry {
+				return "", ErrMixed
+			}
+			data += foldLine("dn: "+e.Add.DN, fw) + "\n"
+			data += "changetype: add\n"
+			for _, add := range e.Add.Attributes {
+				if len(add.Vals) == 0 {
+					return "", errors.New("changetype 'add' requires non empty value list")
+				}
+				for _, v := range add.Vals {
+					ev, t := encodeValue(v)
+					col := ": "
+					if t {
+						col = ":: "
+					}
+					data += foldLine(add.Type+col+ev, fw) + "\n"
+				}
+			}
+
+		case e.Del != nil:
+			hasChange = true
+			if hasEntry {
+				return "", ErrMixed
+			}
+			data += foldLine("dn: "+e.Del.DN, fw) + "\n"
+			data += "changetype: delete\n"
+
+		case e.Modify != nil:
+			hasChange = true
+			if hasEntry {
+				return "", ErrMixed
+			}
+			data += foldLine("dn: "+e.Modify.DN, fw) + "\n"
+			data += "changetype: modify\n"
+			for _, mod := range e.Modify.AddAttributes {
+				if len(mod.Vals) == 0 {
+					return "", errors.New("changetype 'modify', op 'add' requires non empty value list")
+				}
+
+				data += "add: " + mod.Type + "\n"
+				for _, v := range mod.Vals {
+					ev, t := encodeValue(v)
+					col := ": "
+					if t {
+						col = ":: "
+					}
+					data += foldLine(mod.Type+col+ev, fw) + "\n"
+				}
+				data += "-\n"
+			}
+			for _, mod := range e.Modify.DeleteAttributes {
+				data += "delete: " + mod.Type + "\n"
+				for _, v := range mod.Vals {
+					ev, t := encodeValue(v)
+					col := ": "
+					if t {
+						col = ":: "
+					}
+					data += foldLine(mod.Type+col+ev, fw) + "\n"
+				}
+				data += "-\n"
+			}
+			for _, mod := range e.Modify.ReplaceAttributes {
+				if len(mod.Vals) == 0 {
+					return "", errors.New("changetype 'modify', op 'replace' requires non empty value list")
+				}
+				data += "replace: " + mod.Type + "\n"
+				for _, v := range mod.Vals {
+					ev, t := encodeValue(v)
+					col := ": "
+					if t {
+						col = ":: "
+					}
+					data += foldLine(mod.Type+col+ev, fw) + "\n"
+				}
+				data += "-\n"
+			}
+
+			/*
```

Comment:
Is this block needed?  If it doesn't work, are we trying to preserve something in this commented code that requires leaving it in?  I think it's better to remove commented code.